<a href="https://colab.research.google.com/github/VintaBytes/Ciencia-de-datos/blob/main/Ciencia%20De%20Datos%20Con%20Python/CuadernosColab/CienciaDeDatos_Cap%C3%ADtulo027_Primeras_tablas_de_resumen.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Abrir en Colab"/></a>

# Capítulo 27 — Primeras tablas de resumen

## De resultados sueltos a tablas de resumen

En los capítulos anteriores calculamos distintos tipos de resúmenes.

Por un lado, trabajamos con columnas numéricas: promedios, medianas, mínimos, máximos y cuartiles.

Por otro lado, analizamos columnas categóricas contando valores, calculando frecuencias relativas y expresando proporciones como porcentajes.

Ahora vamos a dar un paso más: organizar resultados en tablas de resumen.

Una tabla de resumen no busca mostrar todos los datos originales. Su objetivo es reunir algunos indicadores importantes para responder una pregunta, comparar información o comunicar una primera lectura del dataset.

En este capítulo vamos a construir tablas simples, pensadas para ordenar resultados y facilitar la interpretación.

## Cargar el dataset de trabajo

Vamos a seguir usando el dataset `tips` de Seaborn.

Mantener el mismo dataset nos permite concentrarnos en una idea nueva: cómo organizar resultados en tablas de resumen claras y útiles.

In [13]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("tips")

df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


La tabla muestra las primeras filas del dataset.

Seguimos trabajando con las mismas variables de los capítulos anteriores. En este capítulo no vamos a centrarnos en una columna aislada, sino en cómo reunir distintos resultados en tablas de resumen.

La idea es que una tabla de resumen ayude a leer mejor el dataset sin mostrar todos los registros originales.

## Una primera tabla de indicadores generales

Para comenzar, vamos a construir una tabla con algunos indicadores generales del dataset.

La pregunta que queremos responder es:

**¿Qué información básica podemos resumir sobre las cuentas y las propinas?**

Vamos a calcular la cantidad de cuentas, el total facturado, el total de propinas, el promedio de cuenta y el promedio de propina.

In [14]:
resumen_general = pd.DataFrame({
    "indicador": [
        "Cantidad de cuentas",
        "Total facturado",
        "Total de propinas",
        "Promedio de cuenta",
        "Promedio de propina"
    ],
    "valor": [
        df["total_bill"].count(),
        df["total_bill"].sum(),
        df["tip"].sum(),
        df["total_bill"].mean(),
        df["tip"].mean()
    ]
})

resumen_general["valor"] = resumen_general["valor"].round(2)

resumen_general

,indicador,valor
0,Cantidad de cuentas,244.00
1,Total facturado,4827.77
2,Total de propinas,731.58
3,Promedio de cuenta,19.79
4,Promedio de propina,3.00


La tabla reúne varios indicadores generales del dataset.

En lugar de mostrar cada cálculo por separado, organiza los resultados en una forma más clara:

- hay **244 cuentas** registradas;
- el total facturado es **4827.77**;
- el total de propinas es **731.58**;
- el promedio de cuenta es **19.79**;
- el promedio de propina es **3.00**.

Esta tabla funciona como un primer resumen general. No muestra diferencias por día, horario o tipo de mesa, pero permite comunicar rápidamente algunos datos básicos del conjunto.

## Una tabla de frecuencias

No todas las tablas de resumen se construyen con promedios o sumas.

Cuando trabajamos con columnas categóricas, una tabla de resumen puede mostrar cuántas veces aparece cada categoría y qué porcentaje representa.

Vamos a construir una tabla de frecuencias para la columna `day`.

In [15]:
tabla_dias = pd.DataFrame({
    "frecuencia": df["day"].value_counts(),
    "porcentaje": df["day"].value_counts(normalize=True) * 100
})

tabla_dias["porcentaje"] = tabla_dias["porcentaje"].round(2)

tabla_dias

,frecuencia,porcentaje
day,,
Sat,87,35.66
Sun,76,31.15
Thur,62,25.41
Fri,19,7.79


Esta tabla resume la distribución de la columna `day`.

A diferencia de la tabla anterior, acá cada fila no representa un indicador general, sino una categoría de la columna analizada.

Podemos observar que:

- `Sat` concentra **87 registros**, equivalentes al **35.66%** del total.
- `Sun` concentra **76 registros**, equivalentes al **31.15%**.
- `Thur` concentra **62 registros**, equivalentes al **25.41%**.
- `Fri` concentra **19 registros**, equivalentes al **7.79%**.

Esta tabla permite describir cómo se distribuyen los registros según el día.

## La forma de la tabla depende de la pregunta

Las dos tablas anteriores resumen información, pero no tienen la misma estructura.

En la primera tabla, cada fila era un indicador general del dataset: cantidad de cuentas, total facturado o promedio de propina.

En la segunda tabla, cada fila era una categoría de una columna: `Sat`, `Sun`, `Thur` o `Fri`.

Esto muestra una idea importante: no existe una única forma de construir tablas de resumen. La forma depende de qué pregunta queremos responder y de qué información queremos comunicar.

## Resumir características generales del dataset

También podemos construir una tabla que no se limite solo a indicadores numéricos.

Por ejemplo, podemos resumir cuántos registros tiene el dataset, cuántos días distintos aparecen, cuántos horarios distintos hay registrados y cuáles son los valores promedio de cuenta y propina.

Esta tabla no reemplaza el análisis detallado, pero puede servir como una presentación general del conjunto de datos.

In [16]:
resumen_dataset = pd.DataFrame({
    "aspecto": [
        "Cantidad de registros",
        "Cantidad de días distintos",
        "Cantidad de horarios distintos",
        "Promedio de cuenta",
        "Promedio de propina"
    ],
    "valor": [
        df.shape[0],
        df["day"].nunique(),
        df["time"].nunique(),
        df["total_bill"].mean(),
        df["tip"].mean()
    ]
})

resumen_dataset["valor"] = resumen_dataset["valor"].round(2)

resumen_dataset

,aspecto,valor
0,Cantidad de registros,244.00
1,Cantidad de días distintos,4.00
2,Cantidad de horarios distintos,2.00
3,Promedio de cuenta,19.79
4,Promedio de propina,3.00


Esta tabla reúne características generales del dataset en una sola vista.

Podemos leer que el conjunto tiene **244 registros**, distribuidos en **4 días distintos** y **2 horarios distintos**. También recupera dos indicadores numéricos generales: el promedio de cuenta, **19.79**, y el promedio de propina, **3.00**.

Este tipo de tabla puede ser útil como presentación inicial de un dataset, porque combina información estructural con algunos indicadores básicos.

Sin embargo, todavía no muestra cómo cambian esos indicadores entre categorías. Por ejemplo, no nos dice si el promedio de cuenta es distinto según el día o según el horario.

## Tablas para explorar y tablas para comunicar

Cuando analizamos datos, no todas las tablas cumplen la misma función.

Algunas tablas sirven para explorar. Son tablas de trabajo: pueden tener muchos valores, nombres técnicos o detalles que usamos mientras investigamos el dataset.

Otras tablas están pensadas para comunicar. En ese caso conviene que sean más claras, más breves y que incluyan solo los datos necesarios para responder una pregunta.

La misma información puede organizarse de distintas maneras según el objetivo.

## Preparar una tabla más legible

La tabla anterior es correcta, pero podemos hacerla un poco más clara para presentarla.

Vamos a construir una nueva versión con nombres más descriptivos y algunos valores redondeados.

In [17]:
tabla_presentacion = pd.DataFrame({
    "Resumen": [
        "Registros analizados",
        "Días incluidos",
        "Horarios incluidos",
        "Cuenta promedio",
        "Propina promedio"
    ],
    "Valor": [
        df.shape[0],
        df["day"].nunique(),
        df["time"].nunique(),
        round(df["total_bill"].mean(), 2),
        round(df["tip"].mean(), 2)
    ]
})

tabla_presentacion

,Resumen,Valor
0,Registros analizados,244.00
1,Días incluidos,4.00
2,Horarios incluidos,2.00
3,Cuenta promedio,19.79
4,Propina promedio,3.00


La nueva tabla contiene la misma información general, pero con nombres más pensados para la lectura.

Los nombres de las columnas empiezan con mayúscula y los indicadores usan expresiones más claras, como `Registros analizados`, `Cuenta promedio` y `Propina promedio`.

Este tipo de ajuste puede parecer menor, pero es importante cuando una tabla deja de ser solo una salida de trabajo y empieza a funcionar como parte de una comunicación de resultados.

## Cuidar el formato de los valores

Una tabla puede ser correcta desde el punto de vista del cálculo, pero todavía puede necesitar pequeños ajustes de presentación.

En este caso, algunos valores representan cantidades enteras, como la cantidad de registros, días u horarios. No tiene mucho sentido mostrarlos con decimales.

Vamos a preparar una versión final mostrando las cantidades como enteros y los promedios con dos decimales.

In [18]:
tabla_presentacion = pd.DataFrame({
    "Resumen": [
        "Registros analizados",
        "Días incluidos",
        "Horarios incluidos",
        "Cuenta promedio",
        "Propina promedio"
    ],
    "Valor": [
        df.shape[0],
        df["day"].nunique(),
        df["time"].nunique(),
        f'{df["total_bill"].mean():.2f}',
        f'{df["tip"].mean():.2f}'
    ]
})

tabla_presentacion

,Resumen,Valor
0,Registros analizados,244
1,Días incluidos,4
2,Horarios incluidos,2
3,Cuenta promedio,19.79
4,Propina promedio,3.00


Esta versión queda más cuidada para comunicar resultados.

Los valores que representan cantidades aparecen como números enteros: **244** registros, **4** días y **2** horarios.

En cambio, los valores que representan promedios conservan dos decimales: **19.79** para la cuenta promedio y **3.00** para la propina promedio.

Este detalle mejora la lectura de la tabla. No se trata solo de calcular correctamente, sino también de presentar cada resultado con un formato adecuado.

## Cierre del capítulo

En este capítulo trabajamos con tablas de resumen.

Vimos que una tabla de resumen no tiene una única forma. Puede organizar indicadores generales, frecuencias por categoría o características generales de un dataset.

También distinguimos entre tablas de trabajo y tablas pensadas para comunicar resultados. En una tabla de trabajo puede ser suficiente mostrar la salida tal como la produce Pandas. En una tabla de presentación, en cambio, conviene cuidar los nombres, el orden de los indicadores y el formato de los valores.

Hasta ahora construimos resúmenes generales del dataset o tablas basadas en conteos de categorías. Pero todavía no respondimos preguntas como:

- ¿cuál es el promedio de cuenta por día?
- ¿cuál es la propina promedio según el horario?
- ¿cuántos registros hay en cada combinación de día y horario?

Para responder ese tipo de preguntas vamos a necesitar resumir datos por grupos. Ese será el próximo paso.